### ISSUE TITLE
Noexcept for hashing function objects

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#1480]
I've never really used noexcept, but apparently it's important for hashing functions:

http://stackoverflow.com/questions/13580823/sigfpe-when-accessing-unordered-map

I hope someone can look into this.
Kaldi has various hashing function objects, see stl-utils.h.

**Comment by danpovey** [comment]
@dogancan, do you have time to look at this?

**Comment by dogancan** [comment]
Not at the moment but I can look into it when I have the chance. That might take a while though. From a cursory glance, the idea is to annotate any function that cannot throw with noexcept, right?

**Comment by danpovey** [comment]
not any function, but specifically the () operators of the hashing-function
objects.
apparently there is a performance issue if you don't do that.
Either that, or I thought you might have some insight into whether that was
necessary or not.

**Comment by dogancan** [comment]
Oh yeah that's what I meant to ask. There is some further info in the [GCC manual](https://gcc.gnu.org/onlinedocs/libstdc++/manual/unordered_associative.html) and this [blog post](http://bannalia.blogspot.com/2013/10/implementation-of-c-unordered.html). 

TLDR; C++ standard requires erase and swap operations to be non-throwing. libc++ and libstdc++ implementations of unordered (multi)map/set need access to hash values for these operations. If hash function can throw, it cannot be called by these operations, hence hash values for all items need to be stored along with the items. If hash function is "fast", or if these operations are rare, then storing hash values is wasteful.

**Comment by danpovey** [comment]
OK, we should probably add nothrow then.
There may be hashing functions declared in various places but they should
all contain the string 'Hasher'.

Is there anyone who has time to create a PR for this?  It should be fairly
mechanical.

**Comment by danpovey** [comment]
@tomkocse, do you have time for this?

**Comment by jtrmal** [comment]
If Tom does not have time,  I can look into it.
Y.

**Comment by tomkocse** [comment]
I can start to do this on Monday.

**Comment by danpovey** [comment]
resolved in #1519
